### PR TITLE
[FIX] base: prevent module operations while cron is running

### PR DESCRIPTION
**Pull request by Elkasitu**
Task ID 1960815

### CODE REVIEW

**Comment by mreficent** [comment]
Why not applying the reverse way? I mean, stopping the cron job to do the module operations. When I am going to install/upgrade/uninstall, I don't want to wait.

**Comment by Elkasitu** [comment]
This was already done by @rco-odoo at #20607, this patch is for when it's too late to stop the cron as it has already been executed, there is no choice but to stop the install/upgrade/uninstall

**Comment by mreficent** [comment]
Ah, ok :+1:

**Comment by Elkasitu** [comment]
@Icallhimtest

**Comment by Icallhimtest** [comment]
Nice, thanks for the update!

**Comment by Elkasitu** [comment]
robodoo ~~refry~~ retry

File: odoo/addons/base/module/module.py [hunk 1]
```diff
@@ -479,6 +480,14 @@ def upstream_dependencies(self, known_deps=None,
...
     @api.multi
     def _button_immediate_function(self, function):
+        try:
+            # This is done because the installation/uninstallation/upgrade can modify a currently
+            # running cron job and prevent it from finishing, and since the ir_cron table is locked
+            # during execution, the lock won't be released until timeout.
+            self._cr.execute("SELECT * FROM ir_cron FOR UPDATE NOWAIT")
+        except psycopg2.OperationalError:
+            raise UserError(_("The server is busy right now, module operations are not possible at"
+                              " this time, please try again later."))
```
Comment (rim-odoo) [inline odoo/addons/base/module/module.py]: Hello @Elkasitu 

Please make an error message understandable by the end-user like "The server is currently processing scheduled actions, module installation/upgrade/removal is not possible at this time. Please try again in a few minutes."

I had to grep the code to figure out where this was coming...

By the way, changing that kind of behaviour is quite surprising for a *STABLE* version, even with good intentions.

cc @antonylesuisse @odony

Comment (Elkasitu) [reply odoo/addons/base/module/module.py]: Well it's either that or having a client in deadlock until timeout, which effectively blocks him from working whereas with this patch we temporarily prevent him from installing/upgrading/uninstalling which are not time-critical operations.

Maybe the message could be clearer, not sure if that would've prevented a grep or the user from contacting support though

Comment (rim-odoo) [reply odoo/addons/base/module/module.py]: > Well it's either that or having a client in deadlock until timeout, which effectively blocks him from working whereas with this patch we temporarily prevent him from installing/upgrading/uninstalling which are not time-critical operations.

Which basically occurs... rarely, based on the support history.

> Maybe the message could be clearer, not sure if that would've prevented a grep or the user from contacting support though

Keeping the message "as is" guarantees 100% of confusion...

**Pull request closed without merging**